text/0000-async.md

```diff
@@ -0,0 +1,774 @@
+is constructed here from a boxed future. (In the no_std section, we'll see
+another constructor). In the long run, though, once we can take `dyn` by value,
+we would deprecate `spawn_obj` and add a default `spawn` method:
+
+```rust
+trait Executor {
+    fn spawn(&mut self, task: Async<Output = ()> + Send) -> Result<(), SpawnErrorKind> {
+        self.spawn_obj(TaskObj::new(task))
+    }
+    // ...
+}
+```
+
+At that point we would also deprecate `TaskObj`, which is the reason for using
+the `Obj` suffix -- we want to keep the name `Task` available for potential
+usage down the line.
+
+In addition to the above, the `core::task` module will include the following API
+for helping detect bugs:
+
+```rust
+/// Marks the current thread as being within the dynamic extent of an
+/// executor.
+///
+/// Executor implementations should call this function before beginning to
+/// execute a tasks, and drop the returned `Enter` value after completing
+/// task execution:
+///
+/// ```rust
+/// let enter = enter().expect("...");
+/// /* run task */
+/// drop(enter);
+/// ```
+///
+/// Doing so ensures that executors aren't accidentally invoked in a nested fashion.
+/// When that happens, the inner executor can block waiting for an event that can
+/// only be triggered by the outer executor, leading to a deadlock.
+///
+/// # Error
+///
+/// Returns an error if the current thread is already marked, in which case the
+/// caller should panic with a tailored error message.
+pub fn enter() -> Result<Enter, EnterError>
+```
+
+As stated in the doc comment, the expectation is that all executors will wrap
+their task execution within an `enter` to detect inadvertent nesting.
+
+### Task contexts
+
+All tasks are executed with two pieces of contextual information:
+
+- A `Waker` for waking the task up later on.
+- An executor, which is the "default" place to spawn further tasks.
+
+Notably, this list does *not* include task-local data; that can be addressed
+externally, as we'll see in a later section.
+
+The `core::task::Context` type gathers the (stack-rooted) contextual information
+together, and is passed by mutable reference to all polling functions:
+
+```rust
+/// Information about the currently-running task.
+///
+/// Contexts are always tied to the stack, since they are set up specifically
+/// when performing a single `poll` step on a task.
+pub struct Context<'a> { .. }
+
+impl<'a> Context<'a> {
+    pub fn new(waker: &'a Waker, executor: &'a mut Executor) -> Context<'a>
+
+    /// Get the `Waker` associated with the current task.
+    pub fn waker(&self) -> &Waker;
+
+    /// Run an asynchronous computation to completion on the default executor.
+    ///
+    /// # Panics
+    ///
+    /// This method will panic if the default executor is unable to spawn.
+    /// To handle executor errors, use the `executor` method instead.
+    pub fn spawn(&mut self, f: impl Async<Output = ()> + 'static + Send);
```

Comment:
I'm somewhat concerned about this and associated `'static` bounds.  While I understand that currently the expected implementation is based on a thread pool, I don't really like feeling like this is a decision that every person wanting to implement futures should have to deal with.  Given that (as far as I can tell) there's no justification in the text for why this lifetime needs to be `'static`, is there any chance it could either (1) be changed to a generic lifetime (which a suitable executor could define only for `'static`, if needed) or (2) a note could be added explaining why such a pernicious bound is necessary?

Reply:
I think there might be some confusion here. The idea with the default executor is to make it possible for all futures to assume the ability to spawn new tasks, without having to separately arrange executor access for them. Most executors are going to require `'static` and `Send` tasks.

OTOH it's totally possible -- and very important! -- to have executors that support tasks with borrowed or non-`Send` data. You don't have to use the global/default executor when spawning tasks; you just need to arrange for access to these more specialized executors when you need them. The futures 0.2 crate, for example, provides a `LocalPool` executor that can spawn non-`Send` tasks.

Reply:
Is this all entirely separate from the ability to integrate with async! / await!, then?  That is the part I was mostly worried about.